# بسم الله الرحمن الرحيم

In [24]:
from utils import *
import pandas as pd 
from sklearn.model_selection import train_test_split
print(tf.__version__)

2.11.0


## Reading the Dataset, cleaning , preparing it for the training 

In [2]:
df = pd.read_csv("dataset/body_level_classification_train.csv")

In [3]:
df_original= cleaning_data(df).copy()

In [4]:
all_class_train,all_class_test=data_spliting(df_original,1/5)

In [5]:
oversampled= over_sampling(all_class_train,sampling_ratio=0.8)

In [6]:
X_train=oversampled.drop(["Body_Level"],axis=1).to_numpy()
Y_train=oversampled["Body_Level"].to_numpy()
X_test =all_class_test.drop(["Body_Level"],axis=1).to_numpy()
Y_test =all_class_test["Body_Level"].to_numpy()

In [7]:
X_train, scaler= data_scaling(X_train)
with open(f"scaler","wb") as fp : 
    pickle.dump(scaler,fp)

In [8]:
X_val, X_test, Y_val, Y_test = train_test_split(X_test
                                                     , Y_test, test_size=0.5, random_state=42)

In [9]:
X_val = scaler.transform(X_val)
X_test=scaler.transform(X_test)

In [10]:
model=build_model(2,0.01)

2023-05-18 10:28:12.445380: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory
2023-05-18 10:28:12.445413: W tensorflow/compiler/xla/stream_executor/cuda/cuda_driver.cc:265] failed call to cuInit: UNKNOWN ERROR (303)
2023-05-18 10:28:12.445444: I tensorflow/compiler/xla/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (brownboyComputer): /proc/driver/nvidia/version does not exist
2023-05-18 10:28:12.445840: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.


In [11]:
history = model.fit(X_train,Y_train,validation_data=[X_val,Y_val] 
             ,batch_size=32, epochs=400 , verbose =1, 
                       callbacks=[CustomCallback("NNModels/after_sampling/fifth_model.h5")])

Epoch 1/400
46/46 [==============================] - 2s 7ms/step - loss: 0.8538 - accuracy: 0.4844 - val_loss: 0.4695 - val_accuracy: 0.6463
Epoch 2/400
46/46 [==============================] - 0s 3ms/step - loss: 0.3675 - accuracy: 0.6810 - val_loss: 0.3056 - val_accuracy: 0.7211
Epoch 3/400
46/46 [==============================] - 0s 3ms/step - loss: 0.2743 - accuracy: 0.7449 - val_loss: 0.2425 - val_accuracy: 0.8095
Epoch 4/400
46/46 [==============================] - 0s 3ms/step - loss: 0.2406 - accuracy: 0.7912 - val_loss: 0.2148 - val_accuracy: 0.8299
Epoch 5/400
46/46 [==============================] - 0s 2ms/step - loss: 0.2173 - accuracy: 0.8129 - val_loss: 0.1893 - val_accuracy: 0.8844
Epoch 6/400
46/46 [==============================] - 0s 3ms/step - loss: 0.1986 - accuracy: 0.8408 - val_loss: 0.1836 - val_accuracy: 0.8707
Epoch 7/400
46/46 [==============================] - 0s 3ms/step - loss: 0.1849 - accuracy: 0.8483 - val_loss: 0.1680 - val_accuracy: 0.8912
Epoch 8/400
4

Epoch 56/400
30/46 [==================>...........] - ETA: 0s - loss: 0.0440 - accuracy: 0.9677
Yes You are here 0.9721088409423828 0.9727891087532043
46/46 [==============================] - 0s 3ms/step - loss: 0.0424 - accuracy: 0.9721 - val_loss: 0.0956 - val_accuracy: 0.9728
Epoch 57/400
33/46 [====================>.........] - ETA: 0s - loss: 0.0428 - accuracy: 0.9621
Yes You are here 0.9646258354187012 0.9727891087532043
46/46 [==============================] - 0s 3ms/step - loss: 0.0426 - accuracy: 0.9646 - val_loss: 0.0952 - val_accuracy: 0.9728
Epoch 58/400
46/46 [==============================] - 0s 2ms/step - loss: 0.0412 - accuracy: 0.9660 - val_loss: 0.0969 - val_accuracy: 0.9796
Epoch 59/400
46/46 [==============================] - 0s 2ms/step - loss: 0.0403 - accuracy: 0.9660 - val_loss: 0.0985 - val_accuracy: 0.9796
Epoch 60/400
46/46 [==============================] - 0s 2ms/step - loss: 0.0401 - accuracy: 0.9680 - val_loss: 0.0966 - val_accuracy: 0.9796
Epoch 61/400
3

31/46 [===================>..........] - ETA: 0s - loss: 0.0301 - accuracy: 0.9768
Yes You are here 0.9789115786552429 0.9795918464660645
46/46 [==============================] - 0s 3ms/step - loss: 0.0294 - accuracy: 0.9789 - val_loss: 0.1026 - val_accuracy: 0.9796
Epoch 89/400
29/46 [=================>............] - ETA: 0s - loss: 0.0280 - accuracy: 0.9752
Yes You are here 0.974829912185669 0.9795918464660645
46/46 [==============================] - 0s 3ms/step - loss: 0.0282 - accuracy: 0.9748 - val_loss: 0.1051 - val_accuracy: 0.9796
Epoch 90/400
30/46 [==================>...........] - ETA: 0s - loss: 0.0275 - accuracy: 0.9771
Yes You are here 0.9755101799964905 0.9795918464660645
46/46 [==============================] - 0s 3ms/step - loss: 0.0275 - accuracy: 0.9755 - val_loss: 0.1037 - val_accuracy: 0.9796
Epoch 91/400
32/46 [===================>..........] - ETA: 0s - loss: 0.0279 - accuracy: 0.9746
Yes You are here 0.9734693765640259 0.9795918464660645
46/46 [================

31/46 [===================>..........] - ETA: 0s - loss: 0.0211 - accuracy: 0.9849
Yes You are here 0.9809523820877075 0.9863945841789246
46/46 [==============================] - 0s 3ms/step - loss: 0.0219 - accuracy: 0.9810 - val_loss: 0.1107 - val_accuracy: 0.9864
Epoch 119/400
46/46 [==============================] - 0s 3ms/step - loss: 0.0217 - accuracy: 0.9816 - val_loss: 0.1077 - val_accuracy: 0.9796
Epoch 120/400
46/46 [==============================] - 0s 2ms/step - loss: 0.0212 - accuracy: 0.9837 - val_loss: 0.1050 - val_accuracy: 0.9796
Epoch 121/400
46/46 [==============================] - 0s 2ms/step - loss: 0.0215 - accuracy: 0.9816 - val_loss: 0.1064 - val_accuracy: 0.9796
Epoch 122/400
46/46 [==============================] - 0s 3ms/step - loss: 0.0208 - accuracy: 0.9823 - val_loss: 0.1096 - val_accuracy: 0.9796
Epoch 123/400
46/46 [==============================] - 0s 3ms/step - loss: 0.0211 - accuracy: 0.9816 - val_loss: 0.1085 - val_accuracy: 0.9796
Epoch 124/400
46/4

Epoch 167/400
46/46 [==============================] - 0s 2ms/step - loss: 0.0168 - accuracy: 0.9850 - val_loss: 0.1078 - val_accuracy: 0.9796
Epoch 168/400
32/46 [===================>..........] - ETA: 0s - loss: 0.0146 - accuracy: 0.9893
Yes You are here 0.9843537211418152 0.9863945841789246
46/46 [==============================] - 0s 3ms/step - loss: 0.0158 - accuracy: 0.9844 - val_loss: 0.1118 - val_accuracy: 0.9864
Epoch 169/400
31/46 [===================>..........] - ETA: 0s - loss: 0.0165 - accuracy: 0.9839
Yes You are here 0.9843537211418152 0.9863945841789246
46/46 [==============================] - 0s 3ms/step - loss: 0.0156 - accuracy: 0.9844 - val_loss: 0.1132 - val_accuracy: 0.9864
Epoch 170/400
26/46 [===============>..............] - ETA: 0s - loss: 0.0159 - accuracy: 0.9820
Yes You are here 0.9829931855201721 0.9863945841789246
46/46 [==============================] - 0s 3ms/step - loss: 0.0165 - accuracy: 0.9830 - val_loss: 0.1146 - val_accuracy: 0.9864
Epoch 171/400


Epoch 211/400
43/46 [===========================>..] - ETA: 0s - loss: 0.0132 - accuracy: 0.9862
Yes You are here 0.9857142567634583 0.9863945841789246
46/46 [==============================] - 0s 2ms/step - loss: 0.0135 - accuracy: 0.9857 - val_loss: 0.1104 - val_accuracy: 0.9864
Epoch 212/400
43/46 [===========================>..] - ETA: 0s - loss: 0.0136 - accuracy: 0.9855
Yes You are here 0.9850339889526367 0.9863945841789246
46/46 [==============================] - 0s 2ms/step - loss: 0.0142 - accuracy: 0.9850 - val_loss: 0.1094 - val_accuracy: 0.9864
Epoch 213/400
46/46 [==============================] - 0s 2ms/step - loss: 0.0130 - accuracy: 0.9884 - val_loss: 0.1094 - val_accuracy: 0.9796
Epoch 214/400
46/46 [==============================] - 0s 3ms/step - loss: 0.0139 - accuracy: 0.9871 - val_loss: 0.1116 - val_accuracy: 0.9796
Epoch 215/400
46/46 [==============================] - 0s 3ms/step - loss: 0.0137 - accuracy: 0.9864 - val_loss: 0.1098 - val_accuracy: 0.9864
Epoch 216

46/46 [==============================] - 0s 3ms/step - loss: 0.0120 - accuracy: 0.9871 - val_loss: 0.1095 - val_accuracy: 0.9864
Epoch 250/400
32/46 [===================>..........] - ETA: 0s - loss: 0.0113 - accuracy: 0.9922
Yes You are here 0.9884353876113892 0.9863945841789246
46/46 [==============================] - 0s 3ms/step - loss: 0.0127 - accuracy: 0.9884 - val_loss: 0.1084 - val_accuracy: 0.9864
Epoch 251/400
28/46 [=================>............] - ETA: 0s - loss: 0.0110 - accuracy: 0.9944
Yes You are here 0.9897959232330322 0.9863945841789246
46/46 [==============================] - 0s 3ms/step - loss: 0.0120 - accuracy: 0.9898 - val_loss: 0.1030 - val_accuracy: 0.9864
Epoch 252/400
32/46 [===================>..........] - ETA: 0s - loss: 0.0113 - accuracy: 0.9902
Yes You are here 0.9884353876113892 0.9863945841789246
46/46 [==============================] - 0s 3ms/step - loss: 0.0122 - accuracy: 0.9884 - val_loss: 0.1129 - val_accuracy: 0.9864
Epoch 253/400
46/46 [=======

Epoch 282/400
28/46 [=================>............] - ETA: 0s - loss: 0.0121 - accuracy: 0.9866
Yes You are here 0.9870748519897461 0.9863945841789246
46/46 [==============================] - 0s 3ms/step - loss: 0.0112 - accuracy: 0.9871 - val_loss: 0.1062 - val_accuracy: 0.9864
Epoch 283/400
29/46 [=================>............] - ETA: 0s - loss: 0.0090 - accuracy: 0.9957
Yes You are here 0.9918367266654968 0.9863945841789246
46/46 [==============================] - 0s 3ms/step - loss: 0.0110 - accuracy: 0.9918 - val_loss: 0.1060 - val_accuracy: 0.9864
Epoch 284/400
46/46 [==============================] - 0s 2ms/step - loss: 0.0110 - accuracy: 0.9864 - val_loss: 0.1055 - val_accuracy: 0.9864
Epoch 285/400
42/46 [==========================>...] - ETA: 0s - loss: 0.0109 - accuracy: 0.9903
Yes You are here 0.9904761910438538 0.9863945841789246
46/46 [==============================] - 0s 4ms/step - loss: 0.0107 - accuracy: 0.9905 - val_loss: 0.1088 - val_accuracy: 0.9864
Epoch 286/400


Epoch 314/400
33/46 [====================>.........] - ETA: 0s - loss: 0.0096 - accuracy: 0.9905
Yes You are here 0.9884353876113892 0.9863945841789246
46/46 [==============================] - 0s 3ms/step - loss: 0.0107 - accuracy: 0.9884 - val_loss: 0.1117 - val_accuracy: 0.9864
Epoch 315/400
29/46 [=================>............] - ETA: 0s - loss: 0.0113 - accuracy: 0.9849
Yes You are here 0.9870748519897461 0.9863945841789246
46/46 [==============================] - 0s 3ms/step - loss: 0.0118 - accuracy: 0.9871 - val_loss: 0.1022 - val_accuracy: 0.9864
Epoch 316/400
31/46 [===================>..........] - ETA: 0s - loss: 0.0106 - accuracy: 0.9879
Yes You are here 0.9891156554222107 0.9863945841789246
46/46 [==============================] - 0s 3ms/step - loss: 0.0105 - accuracy: 0.9891 - val_loss: 0.1052 - val_accuracy: 0.9864
Epoch 317/400
32/46 [===================>..........] - ETA: 0s - loss: 0.0112 - accuracy: 0.9834
Yes You are here 0.9843537211418152 0.9863945841789246
46/46

Epoch 346/400
32/46 [===================>..........] - ETA: 0s - loss: 0.0095 - accuracy: 0.9902
Yes You are here 0.9904761910438538 0.9863945841789246
46/46 [==============================] - 0s 3ms/step - loss: 0.0092 - accuracy: 0.9905 - val_loss: 0.1034 - val_accuracy: 0.9864
Epoch 347/400
28/46 [=================>............] - ETA: 0s - loss: 0.0092 - accuracy: 0.9900
Yes You are here 0.9911564588546753 0.9863945841789246
46/46 [==============================] - 0s 3ms/step - loss: 0.0092 - accuracy: 0.9912 - val_loss: 0.1018 - val_accuracy: 0.9864
Epoch 348/400
30/46 [==================>...........] - ETA: 0s - loss: 0.0098 - accuracy: 0.9875
Yes You are here 0.9897959232330322 0.9863945841789246
46/46 [==============================] - 0s 3ms/step - loss: 0.0097 - accuracy: 0.9898 - val_loss: 0.1058 - val_accuracy: 0.9864
Epoch 349/400
32/46 [===================>..........] - ETA: 0s - loss: 0.0099 - accuracy: 0.9873
Yes You are here 0.9877551198005676 0.9863945841789246
46/46

32/46 [===================>..........] - ETA: 0s - loss: 0.0094 - accuracy: 0.9883
Yes You are here 0.9911564588546753 0.9863945841789246
46/46 [==============================] - 0s 3ms/step - loss: 0.0089 - accuracy: 0.9912 - val_loss: 0.1005 - val_accuracy: 0.9864
Epoch 380/400
27/46 [================>.............] - ETA: 0s - loss: 0.0094 - accuracy: 0.9861
Yes You are here 0.9897959232330322 0.9863945841789246
46/46 [==============================] - 0s 3ms/step - loss: 0.0087 - accuracy: 0.9898 - val_loss: 0.1013 - val_accuracy: 0.9864
Epoch 381/400
30/46 [==================>...........] - ETA: 0s - loss: 0.0079 - accuracy: 0.9958
Yes You are here 0.9938775300979614 0.9863945841789246
46/46 [==============================] - 0s 3ms/step - loss: 0.0085 - accuracy: 0.9939 - val_loss: 0.1005 - val_accuracy: 0.9864
Epoch 382/400
44/46 [===========================>..] - ETA: 0s - loss: 0.0087 - accuracy: 0.9929
Yes You are here 0.9925169944763184 0.9863945841789246
46/46 [============

In [12]:
accurate_train_accuracy,classify_report= test_model(model,X_train,Y_train)

46/46 [==============================] - 0s 1ms/step


In [13]:
accurate_train_accuracy

0.992517006802721

In [14]:
print(classify_report)

              precision    recall  f1-score   support

           0       0.97      1.00      0.99       272
           1       1.00      0.96      0.98       278
           2       0.99      1.00      0.99       376
           3       1.00      1.00      1.00       544

    accuracy                           0.99      1470
   macro avg       0.99      0.99      0.99      1470
weighted avg       0.99      0.99      0.99      1470



In [15]:
accurate_val_accuracy,classify_report= test_model(model,X_val,Y_val)

5/5 [==============================] - 0s 1ms/step


In [16]:
accurate_val_accuracy

0.9863945578231292

In [17]:
print(classify_report)

              precision    recall  f1-score   support

           0       1.00      0.95      0.97        20
           1       0.95      1.00      0.98        20
           2       0.97      1.00      0.99        36
           3       1.00      0.99      0.99        71

    accuracy                           0.99       147
   macro avg       0.98      0.98      0.98       147
weighted avg       0.99      0.99      0.99       147



In [18]:
accurate_test_accuracy,classify_report= test_model(model,X_test,Y_test)

5/5 [==============================] - 0s 2ms/step


In [19]:
accurate_test_accuracy

0.9932432432432432

In [20]:
print(classify_report)

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        18
           1       1.00      1.00      1.00        20
           2       1.00      0.98      0.99        45
           3       0.98      1.00      0.99        65

    accuracy                           0.99       148
   macro avg       1.00      0.99      1.00       148
weighted avg       0.99      0.99      0.99       148



## Notes

- name            : first model 
- split           : 1/3 
- oversampling    : 0.5 
- val_acc         : 0.9836734693877551
- test_acc        : 0.983739837398374
- gamma           : 1 
---------------------------------------------------------------------------------------------
- name            : second model 
- split           : 1/3 
- oversampling    : 0.8
- val_acc         : 0.9877551020408163
- test_acc        : 0.9878048780487805
- gamma           : 1 

---------------------------------------------------------------------------------------------

- name            : forth model 
- split           : 1/5
- oversampling    : 1
- val_acc         : 0.9863945578231292
- test_acc        : 0.9932432432432432
- gamma           : 2

---------------------------------------------------------------------------------------------
- name            : fifth model 
- split           : 1/5
- oversampling    : 0.5
- val_acc         : 0.9863945578231292
- test_acc        : 0.9864864864864865
- gamma           : 2

---------------------------------------------------------------------------------------------



In [21]:
df.columns


Index(['Gender', 'Age', 'Height', 'Weight', 'H_Cal_Consump', 'Veg_Consump',
       'Water_Consump', 'Alcohol_Consump', 'Smoking', 'Meal_Count',
       'Food_Between_Meals', 'Fam_Hist', 'H_Cal_Burn', 'Phys_Act',
       'Time_E_Dev', 'Transport', 'Body_Level'],
      dtype='object')